include/lo2s/cupti/ringbuf.hpp

```diff
@@ -0,0 +1,254 @@
+
+    RingBufDeleter(std::byte* start, size_t size) : start_(start), size_(size)
+    {
+    }
+
+    void operator()(std::byte* p) const
+    {
+        if (p >= start_ && p < start_ + size_)
+        {
+            return;
+        }
+        else
+        {
+            free(p);
+        }
+    }
+
+private:
+    std::byte* start_;
+    size_t size_;
+};
+
+typedef std::unique_ptr<std::byte, RingBufDeleter> rb_ptr;
+
+class RingBufWriter
+{
+public:
+    RingBufWriter(std::string filename, size_t size) : filename_(filename)
+    {
+        fd_ = shm_open(filename.c_str(), O_RDWR, 0600);
+        if (fd_ == -1)
+        {
+            throw std::system_error(errno, std::system_category());
+        }
+
+        ringbuf_ = (struct ringbuf_header*)mmap(NULL, sizeof(struct ringbuf_header) + size,
+                                                PROT_READ | PROT_WRITE, MAP_SHARED, fd_, 0);
+
+        if (ringbuf_ == MAP_FAILED)
+        {
+            throw std::system_error(errno, std::system_category());
+        }
+
+        start_ = (std::byte*)ringbuf_ + sizeof(struct ringbuf_header);
+    }
+
+    std::byte* reserve(size_t size)
+    {
+        if (ringbuf_->fill.load() + size > ringbuf_->size)
+        {
+            return nullptr;
+        }
+
+        if (reserved_ != nullptr)
+        {
+            // TODO: exception here
+            return nullptr;
+        }
+        if (ringbuf_->head.load() + size > ringbuf_->size)
+        {
+            reserved_ = (std::byte*)malloc(size);
+        }
+        else
+        {
+            reserved_ = start_ + ringbuf_->head.load();
+        }
+
+        reserved_size_ = size;
+        return reserved_;
+    }
+
+    bool commit()
+    {
+        if (reserved_ == nullptr)
+        {
+            // TODO: Turn into exception, maybe?
+            return false;
+        }
+
+        if (!(reserved_ >= start_ && reserved_ < start_ + ringbuf_->size))
+        {
+            uint64_t wrap = ringbuf_->size - ringbuf_->head.load();
+            memcpy(start_ + ringbuf_->head.load(), reserved_, wrap);
+            memcpy(start_, reserved_ + wrap, reserved_size_ - wrap);
+
+            free(reserved_);
+        }
+        ringbuf_->head.store((ringbuf_->head.load() + reserved_size_) % ringbuf_->size);
+        ringbuf_->fill.fetch_add(reserved_size_);
+
+        reserved_ = nullptr;
+        reserved_size_ = 0;
+
+        return true;
+    }
+
+private:
+    std::string filename_;
+    std::byte* start_;
+    std::byte* reserved_ = nullptr;
+    size_t reserved_size_ = 0;
+    int fd_;
+    struct ringbuf_header* ringbuf_;
+};
+
+class RingBufReader
+{
+public:
+    RingBufReader(std::string filename, size_t size)
+    {
+        fd_ = shm_open(filename.c_str(), O_CREAT | O_EXCL | O_RDWR, 0600);
+        if (fd_ == -1)
+        {
+            throw std::system_error(errno, std::system_category());
+        }
+
+        ringbuf_ = (struct ringbuf_header*)mmap(NULL, sizeof(struct ringbuf_header) + size,
```

Comment:
There is `EventReader::init_mmap()`, which almost does the same thing, but the right way.

Reply:
We either need `fill` or to leave one byte empty between head and tail, because `head == tail` could mean both full and empty. And I personally prefer fill over leaving one byte free because the latter feels like it would invite off-by-one errors

Reply:
I'd rather "waste" one byte in the buffer than have one atomic 8-byte header thing to take care of.
You take care of off-by-one errors with a proper function to count the free space.

See also https://www.kernel.org/doc/Documentation/circular-buffers.txt

P.S. I do quite frankly wonder why one would not just keep incrementing head/tail indefinitely and only `%` them upon access rather than wrap-around. That would seem to be more elegant to me - and you could use the precious last byte! But I cannot oversee the consequences and would suggest to rather go with the proven approach.